NorthstarDLL/mods/modsavefiles.cpp

```diff
@@ -0,0 +1,532 @@
+				if (fileStr.fail())
+				{
+					spdlog::error("A file was supposed to be loaded but we can't access it?!");
+
+					g_pSquirrel<context>->AsyncCall("NSHandleLoadResult", handle, false, "");
+					mutex.get().unlock();
+					return;
+				}
+
+				std::stringstream stringStream;
+				while (fileStr.peek() != EOF)
+					stringStream << (char)fileStr.get();
+
+				g_pSquirrel<context>->AsyncCall("NSHandleLoadResult", handle, true, stringStream.str());
+
+				fileStr.close();
+				mutex.get().unlock();
+				// side-note: this causes a leak?
+				// when a file is added to the map, it's never removed.
+				// no idea how to fix this - because we have no way to check if there are other threads waiting to use this file(?)
+				// tried to use m.try_lock(), but it's unreliable, it seems.
+			}
+			catch (std::exception ex)
+			{
+				spdlog::error("LOAD FAILED!");
+				mutex.get().unlock();
+				spdlog::error(ex.what());
+			}
+		});
+
+	readThread.detach();
+	return handle;
+}
+
+// Deletes a file asynchronously.
+template <ScriptContext context> void SaveFileManager::DeleteFileAsync(fs::path file)
+{
+	// P.S. I don't like how we have to async delete calls but we do.
+	auto m = std::ref(mutexMap[file]);
+	std::thread deleteThread(
+		[m, file]()
+		{
+			try
+			{
+				m.get().lock();
+
+				fs::remove(file);
+
+				m.get().unlock();
+				// side-note: this causes a leak?
+				// when a file is added to the map, it's never removed.
+				// no idea how to fix this - because we have no way to check if there are other threads waiting to use this file(?)
+				// tried to use m.try_lock(), but it's unreliable, it seems.
+			}
+			catch (std::exception ex)
+			{
+				spdlog::error("DELETE FAILED!");
+				m.get().unlock();
+				spdlog::error(ex.what());
+			}
+		});
+
+	deleteThread.detach();
+}
+
+// Checks if a file contains null characters.
+bool ContainsInvalidChars(std::string str)
+{
+	// we don't allow null characters either, even if they're ASCII characters because idk if people can
+	// use it to circumvent the file extension suffix.
+	return std::any_of(str.begin(), str.end(), [](char c) { return c == '\0'; });
+}
+
+// Checks if the relative path (param) remains inside the mod directory (dir).
+// Paths are restricted to ASCII because encoding is fucked and we decided we won't bother.
+bool IsPathSafe(const std::string param, fs::path dir)
+{
+	auto const normRoot = fs::weakly_canonical(dir);
+	auto const normChild = fs::weakly_canonical(dir / param);
+
+	auto itr = std::search(normChild.begin(), normChild.end(), normRoot.begin(), normRoot.end());
+	// we return if the file is safe (inside the directory) and uses only ASCII chars in the path.
+	return itr == normChild.begin() && std::none_of(
+										   param.begin(),
+										   param.end(),
+										   [](char c)
+										   {
+											   unsigned char unsignedC = static_cast<unsigned char>(c);
+											   return unsignedC > 127 || unsignedC < 0;
```

Comment:
Why not just limit it to alnum specifically instead of ASCII in general? We should also restrict the special files on Windows (see https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file) (as an alternative, we could prepend a common suffix like an underscore to the filenames).

Reply:
This was introduced to avert the encoding problems, not to check if a file name is actually "valid".

There are many, _many_ prerequisites for that and I think implementing all of them would be neigh impossible; The best way to check if a file is valid is to try and write to it. I do think that if a save fails, it shouldn't be silent, however, since we have to write to the file asynchronously, that's what is currently done. Maybe we should add a callback for that?

Reply:
This was introduced to avert the encoding problems, not to check if a file name is actually "valid".

There are many, _many_ prerequisites for that and I think implementing all of them would be neigh impossible; The best way to check if a file is valid is to try and write to it. I do think that if a save fails, it shouldn't be silent, however, since we have to write to the file asynchronously, that's what is currently done. Maybe we should add a callback for that?